### ISSUE TITLE
Web-standard ML model format

### ISSUE THREAD

**Issue by josephrocca** [webmachinelearning/model-loader#15]
In the explainer it says:

> For Web ML APIs to succeed, there will need to be converters from all major ML formats into the web standard representation.

As someone who has been wrestling model converters on and off for a while now, I think this is very important. When trying to get pre-trained Pytorch models working in tfjs, tfjs-tflite or ONNX Runtime Web, I inevitably run into lots of problems with unsupported operators and/or unsupported data types. It can be a bit of a nightmare (at least for a humble web developer like myself) because it requires digging into the model file with Python scripts to try to replace datatypes (e.g. [today's problem](https://github.com/microsoft/onnxruntime/issues/9760)), change op attributes, and so forth.

Having a standard ML format and a robust set of converters from the major Python training frameworks would be *extremely* useful for me - but it would all come down to the robustness (data-type/op coverage) of the converters.

So, what that said, it seems like the core of the work for the model loader proposal is really in this design of this "web-standard ML model format" dependency. I'm wondering whether there has been any discussion/resolution on this?

I noticed that a lot of discussion on various approaches was recently removed from the explainer (@jbingham) - I'm guessing this is either because some clarity/direction has emerged on those discussion points, or the issue is being punted while Chrome works on gauging developer demand/requirements with their tflite prototype. If it's the former, it would be very useful to have explanations in the explainer of which approaches are being taken. I'm especially curious about any developments on the "[What is the right level of abstraction for the model format?](https://github.com/webmachinelearning/model-loader/blob/c06b445aa3de97300e83603ac4a6613cebcaaaf1/explainer.md#what-is-the-right-level-of-abstraction-for-the-model-format)" section, and the "[Is a model loader API blocked on MLIR or an alternative intermediate representation becoming a standard?](https://github.com/webmachinelearning/model-loader/blob/c06b445aa3de97300e83603ac4a6613cebcaaaf1/explainer.md#is-a-model-loader-api-blocked-on-mlir-or-an-alternative-intermediate-representation-becoming-a-standard)" section.

Thanks!

**Comment by jbingham** [comment]
@josephrocca Agree with everything you've written! 

The model format discussion has been postponed. Your guess is correct.

For the "level of abstraction of the model format", in the Web NN group, we've decided to decompose operations into the lowest level that we can, to allow composite operations. This addresses some of the concerns that we had with the NN API and operation explosion in TensorFlow.

For "is a model loader API blocked on MLIR", we've decided that it's not. We're focused on getting developer feedback and gauging interest. We hope to start by confirming demand. We'll also work out some of the security and other challenges in implementing the API.

**Comment by josephrocca** [comment]
@jbingham Thanks for your reply 🙏 Good to hear things are moving along, and I like the idea of pushing forward with a model loader API ahead of any model builder API, since the latter seems like a clear super-set of the former (in terms of functionality for developers and browser implementation considerations), and the former is useful in itself, so it's a great stepping stone.

I'm curious what would indicate developer demand with Chrome's experiments, since the utility here seems predicated on the web standard format alleviating the issues that I mentioned. Or are there other key developer pain-points (i.e. unrelated to format conversion problems) that the model loader aims to solve with its initial tests? Perhaps the performance boost will be significant enough on its own?

I'm excited for this API, and probably overly fearful that the initial tests might not show sufficient demand due to not hitting the main pain points that developers have. But it could also be that my conversion woes are not that common and perhaps are related to my lack of experience with ML. In any case, thanks for entertaining my questions :)

**Comment by jbingham** [comment]
If anyone hasn't encountered difficulties with ML format conversion, it's because they've never tried to convert from one format to another :) It's definitely a challenge!

With a working prototype (and origin trial) in Chrome, web developers will be able to run arbitrary TF Lite models. We'll get feedback on what kinds of models they want to run, what operations are required, and what performance benefits there are in practice. We'll have concrete requirements for the model format. Then we can prioritize implementation, and also justify doing our part in TensorFlow to support conversions.

**Comment by josephrocca** [comment]
Sounds great - can't wait to try it out!

**Comment by josephrocca** [comment]
> With a working prototype (and origin trial) in Chrome, web developers will be able to run arbitrary TF Lite models.

Hey @jbingham, I figured I'd ask here rather than in #28, because #28 seems more related to the standard format than the tflite origin trial. One thing that I wanted to bring up is that I've had some trouble ([1](https://github.com/google/jax/issues/9572), [2](https://github.com/google/jax/issues/9659)) converting JAX models to tflite due to the lack of XLA:

 https://github.com/google/jax/blob/main/jax/experimental/jax2tf/g3doc/no_xla_limitations.md

> However, these ops are only be executable by a consumer that has XLA linked in, and this is not the case for the TF.js and TFLite converter. Therefore we provide limited support for these ops by implementing them in terms of ops that are supported in [impl_no_xla.py](https://github.com/google/jax/blob/main/jax/experimental/jax2tf/impl_no_xla.py).

The main reason I'm excited about the Model Loader API is because it seems like it will make the conversion process much less of a head ache. So I'm wondering (as someone who doesn't know much about what's happening behind the scenes) if there is a pathway here toward being able to easily run JAX models in the browser, hopefully even during the origin trial?

**Comment by jbingham** [comment]
Conversion is definitely a challenge! The Model Loader API doesn't escape from the problem. 

For now, TF Lite flatbuffers are used for prototyping. The eventual model format for the web standard is still TBD. Data scientists or developers will need to convert from their trained model (like JAX) to the future web-standard format. That conversion code will need to work flawlessly for multiple ML frameworks. 

The Model Loader API *could* support multiple model formats itself. That would put the burden of conversion on the browser, and require maintaining and updating the conversion code in every browser on every platform. For now, the plan is that the conversion will happen offline, as the general solution.
